saleor/discount/utils.py

```diff
@@ -1132,6 +1208,299 @@ def update_rule_variant_relation(
+    order: "Order",
+    lines_info: Iterable["DraftOrderLineInfo"],
+    database_connection_name: str = settings.DATABASE_CONNECTION_DEFAULT_NAME,
+):
+    create_order_line_discount_objects_for_catalogue_promotions(lines_info)
+    create_order_discount_objects_for_order_promotions(
+        order, lines_info, database_connection_name=database_connection_name
+    )
+    _copy_unit_discount_data_to_order_line(lines_info)
+
+
+def create_order_line_discount_objects_for_catalogue_promotions(
+    lines_info: Iterable[DraftOrderLineInfo],
+):
+    discount_data = prepare_line_discount_objects_for_catalogue_promotions(lines_info)
+    if not discount_data or not lines_info:
+        return
+
+    (
+        discounts_to_create_inputs,
+        discounts_to_update,
+        discount_to_remove,
+        updated_fields,
+    ) = discount_data
+
+    new_line_discounts = []
+    with allow_writer():
+        with transaction.atomic():
+            # Protect against potential thread race. OrderLine object can have only
+            # single catalogue discount applied.
+            order_id = lines_info[0].line.order_id  # type: ignore[index]
+            _order_lock = list(
+                Order.objects.filter(id=order_id).select_for_update(of=(["self"]))
+            )
+
+            if discount_ids_to_remove := [
+                discount.id for discount in discount_to_remove
+            ]:
+                OrderLineDiscount.objects.filter(id__in=discount_ids_to_remove).delete()
+
+            if discounts_to_create_inputs:
+                new_line_discounts = [
+                    OrderLineDiscount(**input) for input in discounts_to_create_inputs
+                ]
+                OrderLineDiscount.objects.bulk_create(
+                    new_line_discounts, ignore_conflicts=True
+                )
+
+            if discounts_to_update and updated_fields:
+                OrderLineDiscount.objects.bulk_update(
+                    discounts_to_update, updated_fields
+                )
+
+    _update_line_info_cached_discounts(
+        lines_info, new_line_discounts, discounts_to_update, discount_ids_to_remove
+    )
+
+    affected_line_ids = [
+        discount_line.line.id
+        for discount_line in new_line_discounts
+        + discounts_to_update
+        + discount_to_remove
+    ]
+    modified_lines_info = [
+        line_info for line_info in lines_info if line_info.line.id in affected_line_ids
+    ]
+    # base unit price must reflect all actual catalogue discounts
+    _update_base_unit_price_amount(modified_lines_info)
+
+
+def _copy_unit_discount_data_to_order_line(lines_info: Iterable[DraftOrderLineInfo]):
+    for line_info in lines_info:
+        if discounts := line_info.discounts:
+            line = line_info.line
+            discount_amount = sum([discount.amount_value for discount in discounts])
+            unit_discount_amount = discount_amount / line.quantity
+            discount_reason = ";".join(
```

Comment:
Maybe adding a space to join would be good? 
```suggestion
            discount_reason = "; ".join(
```